### DISCUSSION TITLE
Configuring logging with node server

### DISCUSSION THREAD

**Discussion by gee4vee** [eclipse-glsp/glsp#957] [Q&A] [answered]
We are using `GlspServerLauncher` to launch our server as part of our vscode extension. Is there a way for us to configure the logging at startup, e.g. change log levels?

Also, is there a way to redirect logging entirely? Specifically, we would like to be able to redirect the GLSP server logging to the Output view in vscode alongside the logging from our other components.

**Comment by tortmayr** [comment, marked as answer]
Hi @gee4vee,

>We are using GlspServerLauncher to launch our server as part of our vscode extension. Is there a way for us to configure the logging at startup, e.g. change log levels?

The glsp-server supports a couple of configuration options that can be passed during startup:
```ts
Usage: Launch a GLSP server [options]

GLSP server

Options:
  -V, --version               output the version number
  -l , --logLevel <logLevel>  Set the log level. [default='info']
  -d , --logDir <logDir>      Set the directory for log files (when file logging is enabled) [default=.config] 
  -p , --port <port>          Set server port [default= 5007] 
  --host <host>               Set host name [default= localhost]
  --no-consoleLog             Disable console logging
  --fileLog                   Enable file logging (default: false)
  -h, --help                  display help for command
  ```
  
 You can [configure](https://github.com/eclipse-glsp/glsp-examples/blob/dcfd58db1cb6890e5e43c8dbe8f12829a44886c0/project-templates/node-json-vscode/glsp-client/tasklist-vscode/extension/src/tasklist-extension.ts#L33) these argument in your `GlspServerLauncher` as part of the `additionalArgs` in your constructor options.
 
 For the logLevel valid options are: `none`,`error`,`warn`,`info`,`debug`,

 For example:
  ```ts
  export async function activate(context: vscode.ExtensionContext): Promise<void> {
        const serverProcess = new GlspServerLauncher({
            executable: path.join(__dirname, '../../../../glsp-server/lib/index.js'),
            socketConnectionOptions: { port: JSON.parse(process.env.TASKLIST_SERVER_PORT || DEFAULT_SERVER_PORT) },
            additionalArgs: ['--loglevel', `debug`, ...)],
            logging: true,
            serverType: 'node'
        });
        ...
    }
   ```
   
   >Also, is there a way to redirect logging entirely? Specifically, we would like to be able to redirect the GLSP server logging to the Output view in vscode alongside the logging from our other components.
   
  Yes this should be possible. You can override the `handleStdoutData` method of the `GlspServerLauncher` and forward the received lines to an outputchannel. I guess something like this should work:
  
  ```ts
// extension.ts
let outputChannel: vscode.OutputChannel | undefined;

export class MyServerLauncher extends GlspServerLauncher {
    protected override handleStdoutData(data: string | Buffer): void {
        if (outputChannel) {
            outputChannel.appendLine(data.toString());
        }
    }
}


export async function activate(context: vscode.ExtensionContext): Promise<void> {
    // Start server process using quickstart component
    outputChannel = vscode.window.createOutputChannel('tasklist-glsp');
    outputChannel.show();
     context.subscriptions.push(outputChannel);
        const serverProcess = new MyServerLauncher(...);
}
  ```
  You also have to make sure that you don`t disable console logging for the glsp server (don`t set the `--no-consoleLog` flag)

**Reply by gee4vee** [reply]
Perfect thank you!

**Comment by gee4vee** [comment]
@tortmayr Another related question: is the logging level adjustable at runtime or only during server startup?

**Comment by tortmayr** [comment]
AFAIK at the moment the level is only adjustable at startup. I haven't looked into it in details but I'm pretty certain that some minor refactorings are required to support runtime adjustmens.
Feel free to open a [feature request](https://github.com/eclipse-glsp/glsp/issues/new/choose) for this.
